extensions/markdown-language-features/src/lines.ts

```diff
@@ -0,0 +1,49 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as vscode from 'vscode';
+
+
+export function createTextLine(line: number, offset: number, text: string, lineBreak: string): vscode.TextLine {
+	let wsRe = /s*/;
+	let firstNws = 0;
+	if (wsRe.test(text)) {
+		firstNws = wsRe.lastIndex;
+	}
+	return {
+		lineNumber: line,
+		text: text,
+		range: new vscode.Range(line, offset, line, offset + text.length),
+		rangeIncludingLineBreak: new vscode.Range(line, offset, line, offset + text.length + lineBreak.length),
+		firstNonWhitespaceCharacterIndex: firstNws,
+		isEmptyOrWhitespace: firstNws === text.length
+	};
+}
+
+export function toTextLines(text: string): vscode.TextLine[] {
+	let current = 0;
+	let result: vscode.TextLine[] = [];
+	while (true) {
+		let nextNewLine = text.indexOf('\r\n', current);
```

Comment:
Using `indexOf` inside another loop will be rather expensive for documents with long lines. What about `.split` + `.map` instead?

Reply:
It's not expensive. We don't scan the whole document every time. We just go from the current to the next occurence of \r\n. Overall this loop has linear complexity of the length of the document. Do you agree?

Reply:
I understand what you are pointing at, we have indexOf both for \r\n and for \n, but AFAIU, splitting via regex is also not free, and in such a case, there's no way for us to recover separators.

Reply:
It's `O(line count * line length)`. In many cases max line length will be < 80 but still better to avoid when possible

Reply:
It's actually worse than that. If there's no \r in the document, it will go to the end of the document on every line, thus O(n^2) complexity. Thank you for pointing this out. Will write something more efficient here.

Reply:
I fixed with split by regexp. I didn't know that it's so powerful in JS.